pxtpy/converter.ts

```diff
@@ -1244,14 +1244,25 @@ namespace pxt.py {
                     B.mkText(")"),
                     stmts(n.body))
             }
-            unifyTypeOf(n.iter, mkArrayType(typeOf(n.target)))
+
+            let isStringLike = false;
+            if (n.iter.kind == "Name") {
+                const v = lookupVar((n.iter as py.Name).id);
+                isStringLike = !!v && !!v.pyRetType && find(v.pyRetType) == tpString;
+            } else if (n.iter.kind == "Str") {
```

Comment:
Good point, felt a bit weird having to handle it this way. @mmoskal do you happen to know if is there is a function to identify the type an arbitrary expression will evaluate to at this point? I don't see anything right away, or even a way to actually find the type a ``Call`` expr node will return

Reply:
did you try this: https://github.com/microsoft/pxt/blob/master/pxtpy/converter.ts#L760

Now, it may not work the first time, but should be set in the second pass

Reply:
Ah, I had tried that but it was never right when I was debugging / etc, but that's because I was probably always trying it on the first pass..

Since we're unifying the type, we'll have to wait on that till 2nd+ pass then, right? so just

```typescript
if (currIteration > 0) {
    const typeOfTarget = typeOf(n.target);
    unifyTypeOf(n.iter, typeOf(n.iter) == tpString ? typeOfTarget : mkArrayType(typeOfTarget));
}
```

seems to work? Thanks

Reply:
@mmoskal , is there a notion of running these passes until fix point or is it just some number of iterations?

It seems like one way to handle above is to narrow the iter type to "string | array<target_type>" on the first pass and then let other constraints elsewhere (like inferring test() returns string) continue to narrow the type

Reply:
it runs up to 5 passes, as long as there are new type unifications being made (search for `numUnifies`).

We don't support union types in unification. You would generally need to be able to backtrack to handle them, so here be dragons.